### PR TITLE
CDI-627 also include alternative stereotypes in the check

### PR DESCRIPTION
**Pull request by struberg**
This version also includes @Alternative @Stereotypes. Txs to @tremes for the catch

### CODE REVIEW

**Comment by mkouba** [comment]
Looks good but I would change the wording a little bit.

1. a class `T` =>  a bean class; it shouldn't be any class and `T` is imho confusing here
2. class which contains => class which declares
3. `@Alternative @Stereotype` => `@Alternative` stereotype; this wording is already used above and makes sense
4. an alternative `Bean<T>` => an alternative bean; `BeanAttributes.isAlternative()` defines what it means
5.  for the bean class => whose bean class has the specified name

In other words:

* a **bean class exists** which is annotated with `@Alternative` or an `@Alternative` stereotype, or
* a **bean class exists** which declares a producer field or a producer method which is annotated with `@Alternative` or an `@Alternative` stereotype , or
* an **alternative bean exists** whose bean class has the specified name

**Comment by tremes** [comment]
+1 for @mkouba wording:-)

**Comment by manovotn** [comment]
I don't think @mkouba solution is correct. As far as I recall, we intentionally avoided using **`bean class`** and instead went for just **`class`**. The reason is that (correct me if I am wrong) spec doesn't state what *bean class* actually is (even tough it extensively uses the term). Plus. if you actually veto the alternative `bean class`, you are left with plain class.

**Comment by mkouba** [comment]
I see. Well, the phrase *bean class* is also used in [CDI 1.0 - Declaring selected alternatives for a bean archive](http://docs.jboss.org/cdi/spec/1.0/html/injectionelresolution.html#d0e3098). So the original 1.0 wording was not correct either? IMHO the bean class is defined as the result of `javax.enterprise.inject.spi.Bean.getBeanClass()` aka the _bean implementation class_.

**Comment by manovotn** [comment]
Fair enough, though I don't claim it was incorrect. I think we only meant to stick to mere `class` for the sake of clarity.

**Comment by struberg** [comment]
hi martin! 
ad 1: I fear we cannot use 'bean class' as it would again be confusing because sometimes they never become Beans.
ad 2: fine for me, will change
ad 3: also fine
ad 4: no, would stick with Bean. BeanAttributes is only an intermediate result
ad 5: puh, need to dig deeper

**Comment by mkouba** [comment]
> ad 4: no, would stick with Bean. BeanAttributes is only an intermediate result

I don't think we should use `Bean<T>` as we don't use it anywhere in the spec, _alternative bean_ should be enough IMHO. But let's wait what others think.

**Comment by Emily-Jiang** [comment]
LGTM

**Comment by antoinesd** [comment]
Closing the PR. It's included in #342

**Comment by antoinesd** [comment]
Closing this PR in favour of #342
